### DISCUSSION TITLE
VRoid SDK 0.1.0 preview フィードバック: ModelLoader

### DISCUSSION THREAD

**Discussion by xelfia** [pixiv/vroid-sdk-developers#29] [General]
VRoidSDK 0.1.0プレビュー版で気づいたことがありましたので、フィードバックしたいと思います。
フィードバックはこのような形でいいのでしょうか。

# `ModelLoader.LoadVRMFromBinaryAsync`について

いくつかの提案があります。

## [a] 非同期/スレッド

その中の以下の部分の変更について。

```cs
importerContext.LoadAsync().ContinueWith((result) =>
// (中略)
// [ソース] ↓これについて
});
// [変更] ↓ これが必要なように思います。
}, TaskScheduler.FromCurrentSynchronizationContext());
```

`理由` この変更で正しく動作するようになりました。もし違うようであればご指摘ください。

## [b] 動作のオプション化

// ↓ この部分の有効/無効のオプション化が必要に思います。
```cs
result.Result.ShowMeshes();
```

`理由` しばらくの間、位置調整/アプリケーション側でのリギングなしのVRMが表示され、視覚的に適切でない場合があるためです。

## [c] `VRMImporterContext`への引数

その中の`VRMImporterContext`が持つ各オプション引数を、`ModelLoader.LoadVRMFromBinaryAsync`に加えることを希望します。

`理由` 独自のHDRP対応などで必要です。たとえばオプションの一部のみですが、以下のような方法です。

```cs
private static Task<GameObject> LoadVRMFromBinaryAsync(byte[] characterBinary,
    Func<glTF_VRM_extensions, IMaterialDescriptorGenerator> materialGeneratorFunc = null) {
// (中略)
    VRMImporterContext importerContext = new VRMImporterContext(
        vrmData, materialGenerator: materialGeneratorFunc?.Invoke(vrmData.VrmExtension));
```

## 提案[b][c]は、以下のメソッドも同様です。

* `LoadVRMFromBinary`
* `LoadVrmAsync`
* `LoadVrm`

**Comment by Haresaku** [comment]
フィードバックありがとうございます。以下それぞれ提案していただいた点について回答いたします

> [a] 非同期/スレッド

こちらでご指摘いただいた件はバグです。次回のプレビュー版、あるいは本番リリースまでには修正いたします。
ただ、 `TaskScheduler.FromCurrentSynchronizationContext()` を使う方法を提案していただいてますが、
この場合 `LoadVRMFromBinaryAsync()` を呼び出す際のスレッドコンテキストによってメインスレッドで呼ばれるか、あるいは別のスレッドで呼び出されるかは実行する状態によって変わります。
Unityの場合、特に各種APIがメインスレッドで呼ばなければいけない制約がある都合上、予期しないトラブルが発生しうるため単純に `TaskScheduler.FromCurrentSynchronizationContext()` を呼ぶのは避けることを現在検討しています。
こういったトラブルを発生させないため、 `Initialize` のタイミングで `SynchronizationContext.Current` を設定してもらい、そのコンテキストで必ず実行するようにしようと今考えているのですがいかがでしょう


> [b] 動作のオプション化

こちらですが、オプションを追加することでも対応は可能です。
ですが、より柔軟性をあげるため、現状のGameObjectをそのまま返すのではなく、`RuntimeGltfInstance` を成功時のコールバックの引数として渡したほうがいいと考えております。
一方で、より簡易に使うということを考えると `RuntimeGltfInstance` を返すというのはUniVRMのコンテキストを知らなければいけないため単純にモデルの呼び出しをしたいだけという開発者にとっては余計な学習コストを負わせることになります。
そのトレードオフを精査してから最終的な方針を決めたいと考えております。

> [c] VRMImporterContextへの引数

こちらについてですがVRMImporterContextの呼び出しをカスタマイズできるように検討いたします。
しかし、ご提案の内容ですと、VrmImporterContextに与える引数に応じてそれぞれオプションを追加しなければいけなくなり、複雑になってきます。
将来的にUniVRMが拡張されることも考えると `Func<VRMData, VrmImporterContext>` をオプショナル引数にし、 `VrmImporterContext` の作成を開発者に委ねるという方針が良いかと思いますがいかがでしょう

**Reply by xelfia** [reply]
回答ありがとうございます。

## 全体について

提案は、少ない変更で可能な例として示しました。以下のようないくつかの異なる方針が思い浮かびますが、より良い形があればどれでもいいと思います。

(A) 複数のオーバーロードを用意する。
(B) `ModelLoader`をチュートリアル/サンプルとして、必要ならカスタム版の作成を推奨する。
(C) 少ないオーバーロードで済ませるために、多くの引数を持つシグネチャにする。
(D) コールバック。（非同期メソッドだと、どれがいつどのスレッドでと、余計な心配をするかも知れませんが）

## 個別に

### [a]

> `Initialize`のタイミングで`SynchronizationContext.Current`を設定

以下のようなことを考えましたが、見落としがあるでしょうか。

* 現状では`ContinueWith`での処理内容がメインスレッド専用で、これは当面変わらないという前提であれば、
  * ➡ 自動的にメインスレッドに切り替わるように、メインスレッドで呼ばれることが前提している初期化処理（`Initialize`？）内で、自動的に`SynchronizationContext.Current`を取得する。
* 将来Unityの仕様が変わるかもと考えるということであれば、
  * ➡ オプションで`SynchronizationContext`をオーバライド可能にする。

#### `Initialize`で`SynchronizationContext`を固定していいのかどうかについて

ほかの方のケースで`ModelLoader`の非同期メソッドを複数の用法で呼び出すようなことがありそうか、メインスレッド以外のスレッドに戻って、効率的に継続できる処理がありそうか、これはまだ読み切れていません。

### [b]

`RuntimeGltfInstance`が得られるように変更または、(A)として`RuntimeGltfInstance`を返すオーバーロードがあれば、柔軟性があって助かりそうです。

### [c]

`Func<VRMData, VrmImporterContext>`が利用できれば、柔軟性があって助かりそうです。

（間接的な記述が増えると、初見で少し戸惑う部分もあるので、チュートリアル/サンプルで示されれば安心です）

**Reply by nyamadan** [reply]
はじめまして、 [nyamadan](https://github.com/pixiv/vroid-sdk-developers/discussions/3#discussioncomment-1877844) と申します。

> 全体について

ModelLoaderは(B)の方針です。

> (B) ModelLoaderをチュートリアル/サンプルとして、必要ならカスタム版の作成を推奨する。

必要であればカスタマイズしていただき、ご利用いただくことを想定しております。
私はカスタマイズしていただくために、ModelLoaderは必要以上に多機能にする必要はないと考えております。

> [a]
> InitializeのタイミングでSynchronizationContext.Currentを設定

私は ContinueWith の処理内容は当面メインスレッド専用であると考えています。そのため自動で SynchronizationContext.Current を取得する方針を取ることを考えています。
一方で、VRoid SDKのメソッドでは引数としてSynchronizationContext.Current を外部から与える形で実装されていることがあるため、そこに合わせるべきかどうかというところは悩んでおりました。

> InitializeでSynchronizationContextを固定していいのかどうかについて

私はVRMをロードした後は多くの場合、レンダリング処理に移ると考えています。レンダリング関連処理はメインスレッドでのみ実行可能であることが多くあるとみており、メインスレッドで固定すると多くの場合 SynchronizationContext を意識せず実行できます。
また、私も今回のケースにおいてサブスレッドで処理を進めることが有効なケースを今は思いつかないのですが、もし有効なケースがあるとすればModelLoaderを参考にカスタマイズ実装を作っていただいたほうがより効果的なものを作成できると思います。

**Reply by nyamadan** [reply]
頂いたフィードバックをチームで議論し0.1.0正式版としてリリースしました。

> [a] 非同期/スレッド

ModelLoader.InitializeのタイミングでメインスレッドのSynchronizationContextを引数として与えるようにしました。
このSynchronizationContextがLoadVRMFromBinaryAsyncの呼び出し時に使用されます。

次の提案いただいた方法でも修正はできるのですが、ほかのメソッド（例えば OauthProvider.CreateOauthClient ）のようにインスタンス生成時にSynchronizationContextを与える方法に合わせるためにこちらの方法を選択いたしました。

```cs
importerContext.LoadAsync().ContinueWith((result) =>
// (中略)
// [ソース] ↓これについて
});
// [変更] ↓ これが必要なように思います。
}, TaskScheduler.FromCurrentSynchronizationContext());
```

> [c] VRMImporterContextへの引数

対応いたしました。onGenerateVRMImporterContextという引数を追加しました。
たとえば次のようにご使用いただけます。

```cs
ModelLoader.LoadVrm(
	model, // 読み込むモデル
	(gameObject) => {
		// 読み込み完了後のコールバック
	},
	(progress) => {
		// 読み込み中の進捗コールバック
	},
	(error) => {
		// エラー発生時のコールバック
	},
	(VRMData vrmData) => {
		// vrmDataからVRMImporterContextを作成する
		return new VRMImporterContext(vrmData);
	}
);
```

> [b] 動作のオプション化

今回は対応を見送りました。
需要については大変理解できるのですが、showMeshesを呼び出さなければモデルが表示できないというのはUniVRMの挙動を把握していないとわからないためデフォルトで表示するようにしております。
オプションフラグを付けるという案もあったのですが、[c]の対応後に議論を行ったのもあり、これ以上フラグを増やすのはよくないという結論になりました。

ModelLoaderに限ってはカスタマイズして使用いただくことを想定しております。お手数ですがModelLoaderからshowMeshesの処理を削除し、ご利用いただければと思います。
